File: BinarySearch/C++/981TimeBasedKeyValueStore.cpp
/*
    Design time-based key-value structure, multiple vals at diff times

    Hash map, since timestamps are naturally in order, binary search

    Time: O(log n)
    Space: O(n)
*/

class TimeMap{
private:
	// Create the data structure : {key : [[ts1, v1][ts2, v2],...]}
	std::unordered_map<std::string, vector<pair<int, string>>> data;

public:
	TimeMap(){

	}

	void set(string key, string value, int timestamp){
		// Insert the data into the hashmap
		data[key].push_back({timestamp, value});
	}

	string get(string key, int timestamp){
		// If key not in data, return empty string
		if (data.find(key) == data.end()){
			return "";
		}

		// Use binary search to look for the timestamp
		int left = 0;
		int right = data[key].size() - 1;

		while (left <= right){
			int mid = ((right - left) / 2) + left;
			if (timestamp == data[key][mid].first){
				return data[key][mid].second;
			} else if (timestamp > data[key][mid].first){
				left = mid + 1;
			} else {
				right = mid - 1;
			}
			
		}

		// if timestamp not found, return the closest smaller timestamp.
		if (right >= 0){
			return data[key][right].second;
		} else{
			return "";
		}
    }
};

File: BinarySearch/Python/981TimeBasedKeyValueStore.py
class TimeMap:

    def __init__(self):
        self.data = {}
        

    def set(self, key: str, value: str, timestamp: int) -> None:
        if key not in self.data:
            self.data[key] = self.data.get(key, [])
        self.data[key].append([value, timestamp])


    def get(self, key: str, timestamp: int) -> str:
        # Here we use binary search because time stamps are in order
        if key not in self.data:
            return ""
        right = len(self.data[key]) - 1 
        left = 0

        while left <= right:
            mid = ((right - left) // 2) + left
            if timestamp == self.data[key][mid][1]:
                return self.data[key][mid][0]
            elif timestamp > self.data[key][mid][1]:
                left = mid + 1
            elif timestamp < self.data[key][mid][1]:
                right = mid - 1

        # If timestamp not found, retrn the closest smaller timestamp
        if (right >= 0): 
            return self.data[key][right][0]
        else:
            return  ""




            #[a,1][b,2][c,3][d,4]
     
# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)